File: src/osgODE/RigidBody.cpp
/*!
 * @file RigidBody.cpp
 * @author Rocco Martino
 */
/***************************************************************************
 *   Copyright (C) 2010 - 2013 by Rocco Martino                            *
 *   martinorocco@gmail.com                                                *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU Lesser General Public License as        *
 *   published by the Free Software Foundation; either version 2.1 of the  *
 *   License, or (at your option) any later version.                       *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU Lesser General Public License for more details.                   *
 *                                                                         *
 *   You should have received a copy of the GNU Lesser General Public      *
 *   License along with this program; if not, write to the                 *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/

/* ======================================================================= */
/* ....................................................................... */
#include <osgODE/RigidBody>
#include <osgODE/World>
#include <osgODE/StaticWorld>
#include <osgODE/Joint>
#include <osgODE/Notify>
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
/* ....................................................................... */
/* ======================================================================= */




using namespace osgODE ;




/* ======================================================================= */
/* ....................................................................... */
RigidBody::RigidBody(void):
    m_mass_negative(false),
    m_update_interacting_sphere(false)
{
    dWorldID    static_world = StaticWorld::instance()->getODEWorld() ;

    // The body is initially created in the static world
    m_ODE_body = dBodyCreate(static_world) ;

    dBodySetData(m_ODE_body, this) ;


    // Some default parameters...
    setMaxAngularSpeed(FLT_MAX) ;
    setLinearDamping(0.0) ;
    setAngularDamping(0.0) ;
    setLinearDampingThreshold(0.0) ;
    setAngularDampingThreshold(0.0) ;
    setAutoDisableLinearThreshold(0.01) ;
    setAutoDisableAngularThreshold(0.01) ;
    setAutoDisableFlag(false) ;
    setAutoDisableSteps(60) ;
    setAutoDisableTime(1.0) ;
    setAutoDisableAverageSamplesCount(5) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
RigidBody::RigidBody(const RigidBody& other, const osg::CopyOp& copyop):
    Transformable(other, copyop),
    m_mass_negative( other.m_mass_negative ),
    m_update_interacting_sphere(other.m_update_interacting_sphere),
    m_old_linear_velocity( other.m_old_linear_velocity ),
    m_linear_acceleration( other.m_linear_acceleration ),
    m_old_angular_velocity( other.m_old_angular_velocity ),
    m_angular_acceleration( other.m_angular_acceleration )
{
    dWorldID    static_world = StaticWorld::instance()->getODEWorld() ;
    m_ODE_body = dBodyCreate(static_world) ;


    _cloneODEBody(other.m_ODE_body, m_ODE_body) ;


    dBodySetData(m_ODE_body, this) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
RigidBody::~RigidBody(void)
{
    if(m_ODE_body) {
        dBodyDestroy(m_ODE_body) ;
    }
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::setMass(double mass)
{
    setMass(mass, osg::Vec3(1.0, 1.0, 1.0), dSphereClass) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::setMass(double mass, const osg::Vec3& size, int type_class)
{
    PS_DBG2(    "osgODE::RigidBody::setMass(%p, mass=%f, size=[%f, %f, %f], class=%d)",
                this, mass, size.x(), size.y(), size.z(), type_class ) ;



    // check for negative mass
    if( mass <= 0.0 ) {

        mass = fabs( mass ) ;

        m_mass_negative = true ;

        this->setGravityMode(false) ;



    } else {

        m_mass_negative = false ;

    }






    // The kinematic bodies have mass->invMass = 0, the dBodySetMass restore
    // this value. To preserve the body state I store the kinematic flag here
    // and restore it at the end of this function
    bool    kinematic_flag = this->getKinematic() ;



    dMass   m ;

    switch(type_class)
    {

        case dBoxClass:

            dMassSetBoxTotal(&m, mass, size.x(), size.y(), size.z()) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        case dCylinderClass:

            dMassSetCylinderTotal(&m, mass, 3, (size.x() + size.y()) / 4.0, size.z()) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        case dSphereClass:

            dMassSetSphereTotal(&m, mass, (size.x() + size.y() + size.z()) / 6.0) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        case dCapsuleClass:

            dMassSetCapsuleTotal(&m, mass, 3, (size.x() + size.y()) / 4.0, size.z() - (size.x() + size.y()) / 2.0) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        default:
            PS_FATAL("osgODE::RigidBody::setMass(%p, ...): invalid type (%u)", this, type_class) ;
        break ;
    }



    // Restore the kinematic flag
    this->setKinematic( kinematic_flag ) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::setDensity(double density)
{
    // Assume a unit sphere
    setDensity(density, osg::Vec3(1.0, 1.0, 1.0), dSphereClass) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::setDensity(double density, const osg::Vec3& size, int type_class)
{
    PS_DBG2(    "osgODE::RigidBody::setDensity(%p, density=%f, size=[%f, %f, %f], class=%d)",
                this, density, size.x(), size.y(), size.z(), type_class ) ;



    // check for negative mass
    if( density <= 0.0 ) {

        density = fabs( density ) ;

        m_mass_negative = true ;

        this->setGravityMode(false) ;


    } else {

        m_mass_negative = false ;
    }



    // See setMass(mass, size, type_class) above
    bool    kinematic_flag = this->getKinematic() ;



    dMass   m ;

    switch(type_class)
    {

        case dBoxClass:

            dMassSetBox(&m, density, size.x(), size.y(), size.z()) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        case dCylinderClass:

            dMassSetCylinder(&m, density, 3, (size.x() + size.y()) / 4.0, size.z()) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        case dSphereClass:

            dMassSetSphere(&m, density, (size.x() + size.y() + size.z()) / 6.0) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        case dCapsuleClass:

            dMassSetCapsule(&m, density, 3, (size.x() + size.y()) / 4.0, size.z() - (size.x() + size.y()) / 2.0) ;
            dBodySetMass(m_ODE_body, &m) ;

        break ;




        default:
            PS_FATAL("osgODE::RigidBody::setDensity(%p, ...): invalid type (%u)", this, type_class) ;
        break ;
    }



    // Restore the kinematic flag
    this->setKinematic( kinematic_flag ) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
osgODE::RigidBody::addForce(const osg::Vec3& f, const osg::Vec3& p, bool fl, bool pl)
{
    const int  bit_mask = ( (int)(pl) << 1 )  |  (int)(fl) ;

    switch( bit_mask )
    {
        case 0:
            // force_local=false, point_local=false
            dBodyAddForceAtPos(m_ODE_body,
                                    f.x(), f.y(), f.z(),
                                    p.x(), p.y(), p.z()) ;
        break ;



        case 1:
            // force_local=true, point_local=false
            dBodyAddRelForceAtPos(m_ODE_body,
                                    f.x(), f.y(), f.z(),
                                    p.x(), p.y(), p.z()) ;
        break ;



        case 2:
            // force_local=false, point_local=true
            dBodyAddForceAtRelPos(m_ODE_body,
                                    f.x(), f.y(), f.z(),
                                    p.x(), p.y(), p.z()) ;
        break ;



        case 3:
            // force_local=true, point_local=true
            dBodyAddRelForceAtRelPos(m_ODE_body,
                                        f.x(), f.y(), f.z(),
                                        p.x(), p.y(), p.z()) ;
        break ;



        default:
            PS_ASSERT( false ) ;
        break ;
    }
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
osgODE::RigidBody::addTorque(const osg::Vec3& t, bool local)
{
    if( local ) {

        dBodyAddRelTorque(m_ODE_body, t.x(), t.y(), t.z()) ;


    } else {

        dBodyAddTorque(m_ODE_body, t.x(), t.y(), t.z()) ;

    }
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
osg::BoundingBox
RigidBody::getAABB(void) const
{
    const osg::Vec3 center = getPosition() ;

    return osg::BoundingBox(center, center) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
bool
RigidBody::addToWorldInternal(World* world)
{
    PS_DBG2("osgODE::RigidBody::addToWorldInternal(%p, world=%p)", this, world) ;

    if( dBodyGetWorld(m_ODE_body) != world->getODEWorld() ) {


        /*
         * [1] clone the body
         */

        dBodyID     b = dBodyCreate( world->getODEWorld() ) ;

        _cloneODEBody(m_ODE_body, b) ;




        /*
         * [2] Destroy the old body and replace it with the new one
         */
        dBodyDestroy(m_ODE_body) ;

        m_ODE_body = b ;


        // Store the instance pointer
        dBodySetData(m_ODE_body, this) ;



        /* In the Joint;;checkBodies method (invoked by _notifyJoints) a body
         * is valid if it has non-NULL world, but the world of this body,
         * although valid, has not been setted yet. I provide: */
        setWorldInternal(world) ;



        /*
         * [3] Tell the joints that this body is now part of the simulation
         */
        _notifyJoints() ;



        PS_DBG3("osgODE::RigidBody::addToWorldInternal(%p): ODE body: %p", this, m_ODE_body) ;



        return true ;
    }


    return false ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
bool
RigidBody::removeFromWorldInternal(World* world)
{
    PS_DBG2("osgODE::RigidBody::removeFromWorldInternal(%p, world=%p)", this, world) ;

    if( dBodyGetWorld(m_ODE_body) == world->getODEWorld() ) {


        /*
         * [1] clone the body
         */
        dBodyID     b = dBodyCreate( StaticWorld::instance()->getODEWorld() ) ;

        _cloneODEBody(m_ODE_body, b) ;



        /*
         * [2] Destroy the old body and replace it with the new one
         */
        dBodyDestroy(m_ODE_body) ;

        m_ODE_body = b ;


        // This is not really useful
        dBodySetData(m_ODE_body, this) ;



        // as in addToWorldInternal
        setWorldInternal(NULL) ;



        /*
         * [3] Tell the joints that this body is no longer part of the simulation
         */
        _notifyJoints() ;



        return true ;
    }


    return false ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::update(double step_size)
{
    World*  world = getWorld() ;

    PS_ASSERT1( world != NULL ) ;


    if( m_update_interacting_sphere ) {
        _updateInteractingSphere() ;
    }


    if( m_mass_negative ) {
        osg::Vec3   F = world->getGravity() * this->getMass() ;
        this->addForce( F ) ;
    }


    this->Transformable::update(step_size) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::postUpdate(double step_size)
{
    updateTransformInternal() ;

    _computeAcceleration(step_size) ;

    this->Transformable::postUpdate(step_size) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::updateTransformInternal(void)
{
    const dReal*    p = dBodyGetPosition(m_ODE_body) ;
    const dReal*    r = dBodyGetRotation(m_ODE_body) ;


    osg::Matrix     m(  r[0] , r[4] , r[8] , 0.0,
                        r[1] , r[5] , r[9] , 0.0,
                        r[2] , r[6] , r[10], 0.0,
                        p[0] , p[1] , p[2] , 1.0 ) ;



    this->getMatrixTransform()->setMatrix(m) ;
    this->getMatrixTransform()->dirtyBound() ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::_notifyJoints(void)
{
    unsigned int        size = m_joints.size() ;

    for(unsigned int i=0; i<size; i++) {

        Joint*  j = m_joints[i].get() ;

        if( j ) {
            j->checkBodiesInternal() ;

        } else {
            // eliminate the destroyed joints from the list
            m_joints[i--] = m_joints[--size] ;
        }
    }

    m_joints.resize(size) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::_updateInteractingSphere(void)
{
    const osg::BoundingSphere&  bs = getInteractingSphere() ;

    setInteractingSphere(getPosition(), bs.radius()) ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::_computeAcceleration(double step_size)
{
    osg::Vec3   current_linear_velocity = getLinearVelocity() ;
    osg::Vec3   current_angular_velocity = getAngularVelocity() ;

    m_linear_acceleration = (current_linear_velocity - m_old_linear_velocity) / step_size ;
    m_angular_acceleration = (current_angular_velocity - m_old_angular_velocity) / step_size ;

    m_old_linear_velocity = current_linear_velocity ;
    m_old_angular_velocity = current_angular_velocity ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
RigidBody*
RigidBody::asRigidBody(void)
{
    return this ;
}
/* ....................................................................... */
/* ======================================================================= */




/* ======================================================================= */
/* ....................................................................... */
void
RigidBody::_cloneODEBody(dBodyID src, dBodyID dst)
{
    PS_DBG2("osgODE::RigidBody::_cloneODEBody(src=%p, dest=%p)", src, dst) ;

    // position and orientation
    {
        const dReal*    pos = dBodyGetPosition(src) ;
        const dReal*    rot = dBodyGetRotation(src) ;
        const dReal*    l_vel = dBodyGetLinearVel(src) ;
        const dReal*    a_vel = dBodyGetAngularVel(src) ;

        dBodySetPosition(dst, pos[0], pos[1], pos[2]) ;
        dBodySetRotation(dst, rot) ;
        dBodySetLinearVel(dst, l_vel[0], l_vel[1], l_vel[2]) ;
        dBodySetAngularVel(dst, a_vel[0], a_vel[1], a_vel[2]) ;
    }


    // mass and force
    {
        dMass   m ;
        dBodyGetMass(src, &m) ;
        dBodySetMass(dst, &m) ;


        const dReal*    force = dBodyGetForce(src) ;
        const dReal*    torque = dBodyGetTorque(src) ;

        dBodySetForce(dst, force[0], force[1], force[2]) ;
        dBodySetTorque(dst, torque[0], torque[1], torque[2]) ;
    }


    // kinematic state
    {
        if( dBodyIsKinematic(src) ) {
            dBodySetKinematic(dst) ;
        } else {
            dBodySetDynamic(dst) ;
        }
    }


    // automatic enabling and disabling
    {
        dBodySetAutoDisableFlag(                dst, dBodyGetAutoDisableFlag(                   src)) ;
        dBodySetAutoDisableLinearThreshold(     dst, dBodyGetAutoDisableLinearThreshold(        src)) ;
        dBodySetAutoDisableAngularThreshold(    dst, dBodyGetAutoDisableAngularThreshold(       src)) ;
        dBodySetAutoDisableSteps(               dst, dBodyGetAutoDisableSteps(                  src)) ;
        dBodySetAutoDisableTime(                dst, dBodyGetAutoDisableTime(                   src)) ;
        dBodySetAutoDisableAverageSamplesCount( dst, dBodyGetAutoDisableAverageSamplesCount(    src)) ;
    }


    // damping
    {
        dBodySetLinearDamping(              dst,    dBodyGetLinearDamping(              src)) ;
        dBodySetAngularDamping(             dst,    dBodyGetAngularDamping(             src)) ;
        dBodySetLinearDampingThreshold(     dst,    dBodyGetLinearDampingThreshold(     src)) ;
        dBodySetAngularDampingThreshold(    dst,    dBodyGetAngularDampingThreshold(    src)) ;
        dBodySetMaxAngularSpeed(            dst,    dBodyGetMaxAngularSpeed(            src)) ;
    }


    // miscellaneous
    {
        dBodySetData(dst, dBodyGetData(src)) ;

        dBodySetFiniteRotationMode(dst, dBodyGetFiniteRotationMode(src)) ;

        dVector3    fra ;
        dBodyGetFiniteRotationAxis(src, fra) ;
        dBodySetFiniteRotationAxis(dst, fra[0], fra[1], fra[2]) ;

        dBodySetGravityMode(dst, dBodyGetGravityMode(src)) ;
    }


    // enabled
    // This is at the end of the function because some ODE functions modify this flag
    if( dBodyIsEnabled(src) ) {
        dBodyEnable(dst) ;
    } else {
        dBodyDisable(dst) ;
    }
}
/* ....................................................................... */
/* ======================================================================= */

File: BlenderExporter/io_scene_sirengine/RigidBody.py
# -*- coding: iso-8859-1 -*-
# file RigidBody.py
# author Rocco Martino
#
############################################################################
#    Copyright (C) 2012 - 2014 by Rocco Martino                            #
#    martinorocco@gmail.com                                                #
#                                                                          #
#    This program is free software; you can redistribute it and/or modify  #
#    it under the terms of the GNU General Public License as published by  #
#    the Free Software Foundation; either version 2 of the License, or     #
#    (at your option) any later version.                                   #
#                                                                          #
#    This program is distributed in the hope that it will be useful,       #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of        #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
#    GNU General Public License for more details.                          #
#                                                                          #
#    You should have received a copy of the GNU General Public License     #
#    along with this program; if not, write to the                         #
#    Free Software Foundation, Inc.,                                       #
#    59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             #
############################################################################

############################################################################
from . import ODEObject

import bpy

from mathutils import Vector, Quaternion, Euler
############################################################################




############################################################################
# ........................................................................ #
# ........................................................................ #
############################################################################




############################################################################
# ........................................................................ #
class RigidBody(ODEObject.ODEObject):
    """osgODE::RigidBody"""





############################################################################
    UpdateActorBound = False
    ActorBoundRadius = None
    Position = Vector()
    Rotation = Quaternion()
    GravityMode = True
    LinearDamping = 0.0
    AngularDamping = 0.0
    LinearDampingThreshold = 0.0
    AngularDampingThreshold = 0.0
    MaxAngularSpeed = None
    Mass = 1.0
    Kinematic = False
    GyroscopicMode = True
    AutoDisableFlag = False
    AutoDisableLinearThreshold = 0.05
    AutoDisableAngularThreshold = 0.05
    AutoDisableTime = 0.1
    AutoDisableSteps = 6
    AutoDisableAverageSamplesCount = 1
############################################################################









############################################################################
    def __init__(self, data, obj):
        super(RigidBody, self).__init__(data, obj)

        self.UpdateActorBound = False
        self.ActorBoundRadius = None
        self.Position = Vector()
        self.Quaternion = Quaternion()
        self.GravityMode = True
        self.LinearDamping = 0.0
        self.AngularDamping = 0.0
        self.LinearDampingThreshold = 0.0
        self.AngularDampingThreshold = 0.0
        self.MaxAngularSpeed = None
        self.Mass = 1.0
        self.Kinematic = False
        self.GyroscopicMode = True
        self.AutoDisableFlag = False
        self.AutoDisableLinearThreshold = 0.01
        self.AutoDisableAngularThreshold = 0.01
        self.AutoDisableTime = 1.0
        self.AutoDisableSteps = 60
        self.AutoDisableAverageSamplesCount = 5
############################################################################




############################################################################
    def buildGraph(self):
        if not super(RigidBody, self).buildGraph():
            return False

        self.UpdateActorBound = self.Object.game.use_actor

        self.ActorBoundRadius = self.Object.game.radius

        self.Position, self.Quaternion, scale = self.Object.matrix_local.decompose()

        self.Mass = self.Object.game.mass


        self.GravityMode = not (self.Object.game.use_ghost or self.Object.game.physics_type == 'NO_COLLISION')


        self.LinearDamping = self.Object.game.damping / bpy.data.scenes[0].render.fps
        self.AngularDamping = self.Object.game.rotation_damping / bpy.data.scenes[0].render.fps
        self.LinearDampingThreshold = self.Object.game.velocity_min
        self.AngularDampingThreshold = self.Object.game.velocity_min

        if self.Object.game.velocity_max > 0.0:
            self.MaxAngularSpeed = self.Object.game.velocity_max


        if self.Object.game.physics_type == 'STATIC':
            self.Kinematic = True


        self.AutoDisableFlag = not self.Object.game.use_sleep
        self.AutoDisableLinearThreshold = self.Data.Context.scene.game_settings.deactivation_linear_threshold
        self.AutoDisableAngularThreshold = self.Data.Context.scene.game_settings.deactivation_angular_threshold
        self.AutoDisableTime = self.Data.Context.scene.game_settings.deactivation_time
        self.AutoDisableSteps = self.Data.Context.scene.game_settings.deactivation_time * self.Data.Context.scene.game_settings.fps
        self.AutoDisableAverageSamplesCount = self.AutoDisableSteps / 10


        from . import ODETransform
        self.ODETransform = ODETransform.ODETransform(self.Data, self.Object)


        self.Data.RigidBodyCache.set(self.Object, self.UniqueID)
        self.Data.RigidBodyMap[self.Object] = self



        for prop in self.Object.game.properties:

            prop_type = ""
            prop_name = prop.name
            prop_value = prop.value

            if prop.type == "STRING":
                prop_type = "String"

            elif prop.type == "FLOAT":
                prop_type = "Float"

            elif prop.type == "INT":
                prop_type = "Int"

            elif prop.type == "BOOL":
                prop_type = "Bool"

                if prop.value:
                    prop_value = "TRUE"
                else:
                    prop_value = "FALSE"


            self.addUserValue( prop_name, prop_value, prop_type )



        try:
            if self.Object["oo_camera_path"]:
                self.addUserValue( "oo_camera_path", 1 )
        except:
            pass

        try:
            v = self.Object["oo_sliding_door"]
            self.addUserValue( "oo_sliding_door", v )
        except:
            pass



        if self.Object.animation_data and self.Object.animation_data.action:

            from . import MotionPath

            mp = MotionPath.MotionPath( self.Data, self.Object )

            mp.buildGraph()

            self.addUpdateCallback( mp )




        return self.ODETransform.buildGraph()
############################################################################




############################################################################
    def writeToStream(self, writer):

        writer.moveIn("osgODE::RigidBody") ;

        if not self.writePrivateData(writer):
            return False

        writer.moveOut("osgODE::RigidBody")

        return True
############################################################################




############################################################################
    def writePrivateData(self, writer):

        if not super(RigidBody, self).writePrivateData(writer) :
            return False



        if self.UpdateActorBound :
            writer.writeLine("ActorBound %f %f %f %f" %(self.Position.x, self.Position.y, self.Position.z, self.ActorBoundRadius))



        if self.ODETransform:
            writer.moveIn("MatrixTransform TRUE")

            if not self.ODETransform.writeToStream(writer):
                return False

            writer.moveOut("MatrixTransform TRUE")



        writer.writeLine("Position %f %f %f" %(self.Position.x, self.Position.y, self.Position.z))
        writer.writeLine("Quaternion %f %f %f %f" %(self.Quaternion[1], self.Quaternion[2], self.Quaternion[3], self.Quaternion[0]))

        if not self.GravityMode:
            writer.writeLine("GravityMode FALSE")



        if self.AutoDisableFlag:
            writer.writeLine("AutoDisableFlag TRUE")
        else:
            writer.writeLine("AutoDisableFlag FALSE")

        writer.writeLine("AutoDisableLinearThreshold %f" %self.AutoDisableLinearThreshold)
        writer.writeLine("AutoDisableAngularThreshold %f" %self.AutoDisableAngularThreshold)
        writer.writeLine("AutoDisableSteps %u" %self.AutoDisableSteps)
        writer.writeLine("AutoDisableTime %f" %self.AutoDisableTime)
        writer.writeLine("AutoDisableAverageSamplesCount %i" %self.AutoDisableAverageSamplesCount)



        writer.writeLine("LinearDamping %f" %self.LinearDamping)
        writer.writeLine("AngularDamping %f" %self.AngularDamping)
        writer.writeLine("LinearDampingThreshold %f" %self.LinearDampingThreshold)
        writer.writeLine("AngularDampingThreshold %f" %self.AngularDampingThreshold)

        if self.MaxAngularSpeed != None:
            writer.writeLine("MaxAngularSpeed %f" %self.MaxAngularSpeed)



        writer.writeLine("Mass %f" %self.Mass)



        if self.Kinematic:
            writer.writeLine("Kinematic TRUE")



        if not self.GyroscopicMode:
            writer.writeLine("GyroscopicMode FALSE")


        return True
############################################################################




# ........................................................................ #
############################################################################
